README.md
# A cron parser

Spits out next & previous cron dates starting from now or a specific date.
Please notice that the timezone of the resulting dates are in the
same timezone provided with Cron().parse(...)

## Usage

A simple usage example:
```dart
import 'package:cron_parser/cron_parser.dart';
import 'package:timezone/timezone.dart';

main() {
  // by default next cron dates are starting from the current date
  var cronIterator = Cron().parse("0 * * * *", "Europe/London");
  TZDateTime nextDate = cronIterator.next();
  // you can retrieve the current value by using the current method
  TZDateTime currentDate = cronIterator.current(); // same as nextDate
  TZDateTime afterNextDate = cronIterator.next();

  cronIterator = Cron().parse("0 * * * *", "Europe/London");
  TZDateTime previousDate = cronIterator.previous();
  // you can retrieve the current value by using the current method
  TZDateTime currentDate = cronIterator.current(); // same as previousDate
  TZDateTime beforePreviousDate = cronIterator.previous();
}
```

Another example this time with a specific start date:
```dart
import 'package:cron_parser/cron_parser.dart';
import 'package:timezone/timezone.dart';

main() {
  TZDateTime startDate = TZDateTime(getLocation("Europe/London"), 2020, 4, 01);
  var cronIterator = Cron().parse("0 * * * *", "Europe/London", startDate);
  TZDateTime nextDate = cronIterator.next(); // 2020-04-01 01:00:00.000+0100
  TZDateTime afterNextDate = cronIterator.next(); // 2020-04-01 02:00:00.000+0100

  cronIterator = Cron().parse("0 * * * *", "Europe/London", startDate);
  TZDateTime previousDate = cronIterator.previous(); // 2020-03-31 23:00:00.000+0100
  TZDateTime beforePreviousDate = cronIterator.previous(); // 2020-03-31 22:00:00.000+0100
}
```

## Links

- [source code][source]
- contributors: [Ronny Bubke](https://github.com/rbubke), [Nils Reichardt](https://github.com/nilsreichardt)

[source]: https://github.com/rbubke/cron-parser

CHANGELOG.md
# Changelog

## 0.5.0

- Added previous method, to get the previous crons: `CronIterator().previous()`
- BREAKING CHANGE: `HasNext().current()` moved into `CronIterator().current()`

## 0.4.0

- Migrated to null safety
- Upgraded dependencies

## 0.3.1

- Cron selection definitions now supported


## 0.2.1

- Cron iterator now also makes the current value accessible


## 0.1.5

- Formatted source with dartfmt


## 0.1.4

- Added example


## 0.1.3

- Changed API to support timezones.


## 0.1.2

- Added docs.


## 0.1.1

- Initial version.


lib/cron_test.dart
import 'package:test/test.dart';

import 'cron.dart';

void main() {
  test('Cron().parse() throws exception for invalid cron string', () {
    expect(() => Cron().parse(""), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("*"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* * *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* * * *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("-1 * * * *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("60 * * * *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* -1 * * *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* 24 * * *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* * 0 * *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* * 32 * *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* * * 0 *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* * * 13 *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* * * * -1"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* * * * 8"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("-1-2 * * * *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* -1-2 * * *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* * -1-2 * *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* * * -1-2 *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* * * * -1-2"), throwsA(TypeMatcher<AssertionError>()));
  });

  DateTime normalizedDate([DateTime dateTime = null]) {
    DateTime date = dateTime == null ? DateTime.now() : dateTime;
    return date
        .subtract(Duration(microseconds: date.microsecond, milliseconds: date.millisecond, seconds: date.second));
  }

  group("Cron().parse() delivers", () {
    test('next minutes starting from date', () {
      DateTime date = normalizedDate();
      var cronIterator = Cron().parse("* * * * *");
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 1))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 2))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 3))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 4))));
    });

    test('next hours starting from date', () {
      DateTime date = normalizedDate();
      date = normalizedDate().subtract(Duration(minutes: date.minute));
      var cronIterator = Cron().parse("0 * * * *");
      expect(cronIterator.next(), equals(date.add(Duration(hours: 1))));
      expect(cronIterator.next(), equals(date.add(Duration(hours: 2))));
      expect(cronIterator.next(), equals(date.add(Duration(hours: 3))));
      expect(cronIterator.next(), equals(date.add(Duration(hours: 4))));
    });

    test('next days starting from date', () {
      DateTime date = normalizedDate();
      date = date.subtract(Duration(minutes: date.minute, hours: date.hour));
      var cronIterator = Cron().parse("0 0 * * *");
      expect(cronIterator.next(), equals(date.add(Duration(days: 1))));
      expect(cronIterator.next(), equals(date.add(Duration(days: 2))));
      expect(cronIterator.next(), equals(date.add(Duration(days: 3))));
      expect(cronIterator.next(), equals(date.add(Duration(days: 4))));
    });

    test('next month starting from date', () {
      DateTime date = normalizedDate();
      var cronIterator = Cron().parse("0 0 1 * *");
      expect(cronIterator.next(), equals(DateTime(date.year, date.month + 1, 1)));
      expect(cronIterator.next(), equals(DateTime(date.year, date.month + 2, 1)));
      expect(cronIterator.next(), equals(DateTime(date.year, date.month + 3, 1)));
      expect(cronIterator.next(), equals(DateTime(date.year, date.month + 4, 1)));
      expect(cronIterator.next(), equals(DateTime(date.year, date.month + 5, 1)));
    });

    test('next weekday starting from monday', () {
      DateTime date = normalizedDate();
      date = date.subtract(Duration(minutes: date.minute, hours: date.hour));
      var cronIterator = Cron().parse("0 0 * * 1");
      var previous = cronIterator.next();
      expect(previous.weekday, equals(1));
      var next = cronIterator.next();
      expect(next.weekday, equals(1));
      expect(next, equals(previous.add(Duration(days: 7))));
    });

    test('next weekday starting from sunday', () {
      DateTime date = normalizedDate();
      date = date.subtract(Duration(minutes: date.minute, hours: date.hour));
      var cronIterator = Cron().parse("0 0 * * 0");
      var previous = cronIterator.next();
      expect(previous.weekday, equals(7));
      var next = cronIterator.next();
      expect(next.weekday, equals(7));
      expect(next, equals(previous.add(Duration(days: 7))));
    });

    test('next weekday starting from sunday with 7', () {
      DateTime date = normalizedDate();
      date = date.subtract(Duration(minutes: date.minute, hours: date.hour));
      var cronIterator = Cron().parse("0 0 * * 7");
      var previous = cronIterator.next();
      expect(previous.weekday, equals(7));
      var next = cronIterator.next();
      expect(next.weekday, equals(7));
      expect(next, equals(previous.add(Duration(days: 7))));
    });

    test('next weekday starting from saturday', () {
      DateTime date = normalizedDate();
      date = date.subtract(Duration(minutes: date.minute, hours: date.hour));
      var cronIterator = Cron().parse("0 0 * * 6");
      var previous = cronIterator.next();
      expect(previous.weekday, equals(6));
      var next = cronIterator.next();
      expect(next.weekday, equals(6));
      expect(next, equals(previous.add(Duration(days: 7))));
    });

    test('next weekday starting from friday', () {
      DateTime date = normalizedDate();
      date = date.subtract(Duration(minutes: date.minute, hours: date.hour));
      var cronIterator = Cron().parse("0 0 * * 5");
      var previous = cronIterator.next();
      expect(previous.weekday, equals(5));
      var next = cronIterator.next();
      expect(next.weekday, equals(5));
      expect(next, equals(previous.add(Duration(days: 7))));
    });

    test('next minutes range starting from date', () {
      DateTime date = normalizedDate(DateTime(2020, 03, 31));
      var cronIterator = Cron().parse("0-3 * * * *", date);
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 1))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 2))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 3))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 60))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 61))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 62))));
    });

    test('next hours range starting from date', () {
      DateTime date = normalizedDate(DateTime(2020, 03, 31));
      var cronIterator = Cron().parse("0 0-3 * * *", date);
      expect(cronIterator.next(), equals(date.add(Duration(hours: 1))));
      expect(cronIterator.next(), equals(date.add(Duration(hours: 2))));
      expect(cronIterator.next(), equals(date.add(Duration(hours: 3))));
      expect(cronIterator.next(), equals(date.add(Duration(hours: 24))));
      expect(cronIterator.next(), equals(date.add(Duration(hours: 25))));
      expect(cronIterator.next(), equals(date.add(Duration(hours: 26))));
    });

    test('next days range starting from date', () {
      DateTime date = normalizedDate(DateTime(2020, 03, 29));
      var cronIterator = Cron().parse("0 0 1-3 * *", date);
      expect(cronIterator.next(), equals(DateTime(2020, 04, 1)));
      expect(cronIterator.next(), equals(DateTime(2020, 04, 2)));
      expect(cronIterator.next(), equals(DateTime(2020, 04, 3)));
      expect(cronIterator.next(), equals(DateTime(2020, 05, 1)));
    });

    test('next month range starting from date', () {
      DateTime date = normalizedDate(DateTime(2020, 03, 29));
      var cronIterator = Cron().parse("0 0 4 1-3 *", date);
      expect(cronIterator.next(), equals(DateTime(2021, 01, 4)));
      expect(cronIterator.next(), equals(DateTime(2021, 02, 4)));
      expect(cronIterator.next(), equals(DateTime(2021, 03, 4)));
      expect(cronIterator.next(), equals(DateTime(2022, 01, 4)));
    });

    test('next weekdays range starting from date', () {
      DateTime date = normalizedDate(DateTime(2020, 03, 28));
      var cronIterator = Cron().parse("0 0 * * 0-5", date);
      expect(cronIterator.next(), equals(DateTime(2020, 03, 29)));
      expect(cronIterator.next(), equals(DateTime(2020, 03, 30)));
      expect(cronIterator.next(), equals(DateTime(2020, 03, 31)));
      expect(cronIterator.next(), equals(DateTime(2020, 04, 1)));
      expect(cronIterator.next(), equals(DateTime(2020, 04, 2)));
      expect(cronIterator.next(), equals(DateTime(2020, 04, 3)));
      expect(cronIterator.next(), equals(DateTime(2020, 04, 5)));
      expect(cronIterator.next(), equals(DateTime(2020, 04, 6)));
    });

    test('next minutes interval starting from date', () {
      DateTime date = normalizedDate(DateTime(2020, 03, 28));
      var cronIterator = Cron().parse("*/5 * * * *", date);
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 5))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 10))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 15))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 20))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 25))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 30))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 35))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 40))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 45))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 50))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 55))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 60))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 65))));
    });

    test('next hours interval starting from date', () {
      DateTime date = normalizedDate(DateTime(2020, 03, 28));
      var cronIterator = Cron().parse("0 */2 * * *", date);
      expect(cronIterator.next(), equals(date.add(Duration(hours: 2))));
      expect(cronIterator.next(), equals(date.add(Duration(hours: 4))));
      expect(cronIterator.next(), equals(date.add(Duration(hours: 6))));
    });

    test('next days interval starting from date', () {
      DateTime date = normalizedDate(DateTime(2020, 03, 28));
      var cronIterator = Cron().parse("0 0 */3 * *", date);
      expect(cronIterator.next(), equals(DateTime(2020, 03, 30)));
      expect(cronIterator.next(), equals(DateTime(2020, 04, 3)));
      expect(cronIterator.next(), equals(DateTime(2020, 04, 6)));
    });

    test('next month interval starting from date', () {
      DateTime date = normalizedDate(DateTime(2020, 03, 28));
      var cronIterator = Cron().parse("0 0 1 */2 *", date);
      expect(cronIterator.next(), equals(DateTime(2020, 4, 1)));
      expect(cronIterator.next(), equals(DateTime(2020, 6, 1)));
      expect(cronIterator.next(), equals(DateTime(2020, 8, 1)));
    });

    test('next weekdays interval starting from date', () {
      DateTime date = normalizedDate(DateTime(2020, 03, 28));
      var cronIterator = Cron().parse("0 0 * * */2", date);
      expect(cronIterator.next(), equals(DateTime(2020, 3, 29)));
      expect(cronIterator.next(), equals(DateTime(2020, 3, 31)));
      expect(cronIterator.next(), equals(DateTime(2020, 4, 2)));
      expect(cronIterator.next(), equals(DateTime(2020, 4, 4)));
      expect(cronIterator.next(), equals(DateTime(2020, 4, 5)));
      expect(cronIterator.next(), equals(DateTime(2020, 4, 7)));
    });
  });
}


test/cron_test.dart
import 'package:test/test.dart';

import '../lib/cron.dart';

void main() {
  test('Cron().parse() throws exception for invalid cron string', () {
    expect(() => Cron().parse(""), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("*"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* * *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* * * *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("-1 * * * *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("60 * * * *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* -1 * * *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* 24 * * *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* * 0 * *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* * 32 * *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* * * 0 *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* * * 13 *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* * * * -1"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* * * * 8"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("-1-2 * * * *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* -1-2 * * *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* * -1-2 * *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* * * -1-2 *"), throwsA(TypeMatcher<AssertionError>()));
    expect(() => Cron().parse("* * * * -1-2"), throwsA(TypeMatcher<AssertionError>()));
  });

  DateTime normalizedDate([DateTime dateTime = null]) {
    DateTime date = dateTime == null ? DateTime.now() : dateTime;
    return date
        .subtract(Duration(microseconds: date.microsecond, milliseconds: date.millisecond, seconds: date.second));
  }

  group("Cron().parse() delivers", () {
    test('next minutes starting from date', () {
      DateTime date = normalizedDate();
      var cronIterator = Cron().parse("* * * * *");
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 1))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 2))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 3))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 4))));
    });

    test('next hours starting from date', () {
      DateTime date = normalizedDate();
      date = normalizedDate().subtract(Duration(minutes: date.minute));
      var cronIterator = Cron().parse("0 * * * *");
      expect(cronIterator.next(), equals(date.add(Duration(hours: 1))));
      expect(cronIterator.next(), equals(date.add(Duration(hours: 2))));
      expect(cronIterator.next(), equals(date.add(Duration(hours: 3))));
      expect(cronIterator.next(), equals(date.add(Duration(hours: 4))));
    });

    test('next days starting from date', () {
      DateTime date = normalizedDate();
      date = date.subtract(Duration(minutes: date.minute, hours: date.hour));
      var cronIterator = Cron().parse("0 0 * * *");
      expect(cronIterator.next(), equals(date.add(Duration(days: 1))));
      expect(cronIterator.next(), equals(date.add(Duration(days: 2))));
      expect(cronIterator.next(), equals(date.add(Duration(days: 3))));
      expect(cronIterator.next(), equals(date.add(Duration(days: 4))));
    });

    test('next month starting from date', () {
      DateTime date = normalizedDate();
      var cronIterator = Cron().parse("0 0 1 * *");
      expect(cronIterator.next(), equals(DateTime(date.year, date.month + 1, 1)));
      expect(cronIterator.next(), equals(DateTime(date.year, date.month + 2, 1)));
      expect(cronIterator.next(), equals(DateTime(date.year, date.month + 3, 1)));
      expect(cronIterator.next(), equals(DateTime(date.year, date.month + 4, 1)));
      expect(cronIterator.next(), equals(DateTime(date.year, date.month + 5, 1)));
    });

    test('next weekday starting from monday', () {
      DateTime date = normalizedDate();
      date = date.subtract(Duration(minutes: date.minute, hours: date.hour));
      var cronIterator = Cron().parse("0 0 * * 1");
      var previous = cronIterator.next();
      expect(previous.weekday, equals(1));
      var next = cronIterator.next();
      expect(next.weekday, equals(1));
      expect(next, equals(previous.add(Duration(days: 7))));
    });

    test('next weekday starting from sunday', () {
      DateTime date = normalizedDate();
      date = date.subtract(Duration(minutes: date.minute, hours: date.hour));
      var cronIterator = Cron().parse("0 0 * * 0");
      var previous = cronIterator.next();
      expect(previous.weekday, equals(7));
      var next = cronIterator.next();
      expect(next.weekday, equals(7));
      expect(next, equals(previous.add(Duration(days: 7))));
    });

    test('next weekday starting from sunday with 7', () {
      DateTime date = normalizedDate();
      date = date.subtract(Duration(minutes: date.minute, hours: date.hour));
      var cronIterator = Cron().parse("0 0 * * 7");
      var previous = cronIterator.next();
      expect(previous.weekday, equals(7));
      var next = cronIterator.next();
      expect(next.weekday, equals(7));
      expect(next, equals(previous.add(Duration(days: 7))));
    });

    test('next weekday starting from saturday', () {
      DateTime date = normalizedDate();
      date = date.subtract(Duration(minutes: date.minute, hours: date.hour));
      var cronIterator = Cron().parse("0 0 * * 6");
      var previous = cronIterator.next();
      expect(previous.weekday, equals(6));
      var next = cronIterator.next();
      expect(next.weekday, equals(6));
      expect(next, equals(previous.add(Duration(days: 7))));
    });

    test('next weekday starting from friday', () {
      DateTime date = normalizedDate();
      date = date.subtract(Duration(minutes: date.minute, hours: date.hour));
      var cronIterator = Cron().parse("0 0 * * 5");
      var previous = cronIterator.next();
      expect(previous.weekday, equals(5));
      var next = cronIterator.next();
      expect(next.weekday, equals(5));
      expect(next, equals(previous.add(Duration(days: 7))));
    });

    test('next minutes range starting from date', () {
      DateTime date = normalizedDate(DateTime(2020, 03, 31));
      var cronIterator = Cron().parse("0-3 * * * *", date);
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 1))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 2))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 3))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 60))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 61))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 62))));
    });

    test('next hours range starting from date', () {
      DateTime date = normalizedDate(DateTime(2020, 03, 31));
      var cronIterator = Cron().parse("0 0-3 * * *", date);
      expect(cronIterator.next(), equals(date.add(Duration(hours: 1))));
      expect(cronIterator.next(), equals(date.add(Duration(hours: 2))));
      expect(cronIterator.next(), equals(date.add(Duration(hours: 3))));
      expect(cronIterator.next(), equals(date.add(Duration(hours: 24))));
      expect(cronIterator.next(), equals(date.add(Duration(hours: 25))));
      expect(cronIterator.next(), equals(date.add(Duration(hours: 26))));
    });

    test('next days range starting from date', () {
      DateTime date = normalizedDate(DateTime(2020, 03, 29));
      var cronIterator = Cron().parse("0 0 1-3 * *", date);
      expect(cronIterator.next(), equals(DateTime(2020, 04, 1)));
      expect(cronIterator.next(), equals(DateTime(2020, 04, 2)));
      expect(cronIterator.next(), equals(DateTime(2020, 04, 3)));
      expect(cronIterator.next(), equals(DateTime(2020, 05, 1)));
    });

    test('next month range starting from date', () {
      DateTime date = normalizedDate(DateTime(2020, 03, 29));
      var cronIterator = Cron().parse("0 0 4 1-3 *", date);
      expect(cronIterator.next(), equals(DateTime(2021, 01, 4)));
      expect(cronIterator.next(), equals(DateTime(2021, 02, 4)));
      expect(cronIterator.next(), equals(DateTime(2021, 03, 4)));
      expect(cronIterator.next(), equals(DateTime(2022, 01, 4)));
    });

    test('next weekdays range starting from date', () {
      DateTime date = normalizedDate(DateTime(2020, 03, 28));
      var cronIterator = Cron().parse("0 0 * * 0-5", date);
      expect(cronIterator.next(), equals(DateTime(2020, 03, 29)));
      expect(cronIterator.next(), equals(DateTime(2020, 03, 30)));
      expect(cronIterator.next(), equals(DateTime(2020, 03, 31)));
      expect(cronIterator.next(), equals(DateTime(2020, 04, 1)));
      expect(cronIterator.next(), equals(DateTime(2020, 04, 2)));
      expect(cronIterator.next(), equals(DateTime(2020, 04, 3)));
      expect(cronIterator.next(), equals(DateTime(2020, 04, 5)));
      expect(cronIterator.next(), equals(DateTime(2020, 04, 6)));
    });

    test('next minutes interval starting from date', () {
      DateTime date = normalizedDate(DateTime(2020, 03, 28));
      var cronIterator = Cron().parse("*/5 * * * *", date);
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 5))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 10))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 15))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 20))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 25))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 30))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 35))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 40))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 45))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 50))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 55))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 60))));
      expect(cronIterator.next(), equals(date.add(Duration(minutes: 65))));
    });

    test('next hours interval starting from date', () {
      DateTime date = normalizedDate(DateTime(2020, 03, 28));
      var cronIterator = Cron().parse("0 */2 * * *", date);
      expect(cronIterator.next(), equals(date.add(Duration(hours: 2))));
      expect(cronIterator.next(), equals(date.add(Duration(hours: 4))));
      expect(cronIterator.next(), equals(date.add(Duration(hours: 6))));
    });

    test('next days interval starting from date', () {
      DateTime date = normalizedDate(DateTime(2020, 03, 28));
      var cronIterator = Cron().parse("0 0 */3 * *", date);
      expect(cronIterator.next(), equals(DateTime(2020, 03, 30)));
      expect(cronIterator.next(), equals(DateTime(2020, 04, 3)));
      expect(cronIterator.next(), equals(DateTime(2020, 04, 6)));
    });

    test('next month interval starting from date', () {
      DateTime date = normalizedDate(DateTime(2020, 03, 28));
      var cronIterator = Cron().parse("0 0 1 */2 *", date);
      expect(cronIterator.next(), equals(DateTime(2020, 4, 1)));
      expect(cronIterator.next(), equals(DateTime(2020, 6, 1)));
      expect(cronIterator.next(), equals(DateTime(2020, 8, 1)));
    });

    test('next weekdays interval starting from date', () {
      DateTime date = normalizedDate(DateTime(2020, 03, 28));
      var cronIterator = Cron().parse("0 0 * * */2", date);
      expect(cronIterator.next(), equals(DateTime(2020, 3, 29)));
      expect(cronIterator.next(), equals(DateTime(2020, 3, 31)));
      expect(cronIterator.next(), equals(DateTime(2020, 4, 2)));
      expect(cronIterator.next(), equals(DateTime(2020, 4, 4)));
      expect(cronIterator.next(), equals(DateTime(2020, 4, 5)));
      expect(cronIterator.next(), equals(DateTime(2020, 4, 7)));
    });
  });
}


lib/cron.dart
// Copyright (c) 2020, rbubke. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

import 'dart:async';

abstract class HasNext<E> {
  E next();
}

abstract class Cron {
  factory Cron() => new _Cron();

  HasNext<DateTime> parse(String cronString, [DateTime startTime = null]);
}

class Schedule {
  final List<int> minutes;
  final List<int> hours;
  final List<int> days;
  final List<int> months;
  final List<int> weekdays;

  factory Schedule({dynamic minutes, dynamic hours, dynamic days, dynamic months, dynamic weekdays}) {
    List<int> parsedMinutes = _parseConstraint(minutes)?.where((x) => x >= 0 && x <= 59)?.toList();
    List<int> parsedHours = _parseConstraint(hours)?.where((x) => x >= 0 && x <= 23)?.toList();
    List<int> parsedDays = _parseConstraint(days)?.where((x) => x >= 1 && x <= 31)?.toList();
    List<int> parsedMonths = _parseConstraint(months)?.where((x) => x >= 1 && x <= 12)?.toList();
    List<int> parsedWeekdays =
        _parseConstraint(weekdays)?.where((x) => x >= 0 && x <= 7)?.map((x) => x == 0 ? 7 : x)?.toSet()?.toList();
    return new Schedule._(parsedMinutes, parsedHours, parsedDays, parsedMonths, parsedWeekdays);
  }

  factory Schedule.parse(String cronString) {
    assert(cronString.isNotEmpty);
    String regex0to59 = "([1-5]?[0-9])";
    String regex0to23 = "([1]?[0-9]|[2][0-3])";
    String regex1to31 = "([1-9]|[12][0-9]|[3][01])";
    String regex1to12 = "([1-9]|[1][012])";
    String regex0to7 = "([0-7])";
    String minutesRegex = "($regex0to59([-]$regex0to59)?|[*]([/]$regex0to59)?)";
    String hoursRegex = "($regex0to23([-]($regex0to23))?|[*]([/]$regex0to23)?)";
    String daysRegex = "($regex1to31([-]$regex1to31)?|[*]([/]$regex1to31)?)";
    String monthRegex = "($regex1to12([-]$regex1to12)?|[*]([/]$regex1to12)?)";
    String weekdaysRegex = "($regex0to7([-]$regex0to7)?|[*]([/]$regex0to7)?)";
    assert(
        RegExp("^$minutesRegex\\s+$hoursRegex\\s+$daysRegex\\s+$monthRegex\\s+$weekdaysRegex\$").hasMatch(cronString));
    List<List<int>> p = cronString.split(new RegExp('\\s+')).map(_parseConstraint).toList();
    assert(p.length == 5);
    return new Schedule(minutes: p[0], hours: p[1], days: p[2], months: p[3], weekdays: p[4]);
  }

  Schedule._(this.minutes, this.hours, this.days, this.months, this.weekdays);
}

class _Cron implements Cron {
  @override
  HasNext<DateTime> parse(String cronString, [DateTime startTime]) {
    if (startTime == null) startTime = DateTime.now();
    Schedule schedule = Schedule.parse(cronString);
    return _CronIterator(schedule, startTime);
  }
}

List<int> _parseConstraint(dynamic constraint) {
  if (constraint == null) return null;
  if (constraint is int) return [constraint];
  if (constraint is List<int>) return constraint;
  if (constraint is String) {
    if (constraint == '*') return null;
    final parts = constraint.split(',');
    if (parts.length > 1) {
      final items = parts.map(_parseConstraint).expand((list) => list).toSet().toList();
      items.sort();
      return items;
    }

    int singleValue = int.tryParse(constraint);
    if (singleValue != null) return [singleValue];

    if (constraint.startsWith('*/')) {
      int period = int.tryParse(constraint.substring(2)) ?? -1;
      if (period > 0) {
        return new List.generate(120 ~/ period, (i) => i * period);
      }
    }

    if (constraint.contains('-')) {
      List<String> ranges = constraint.split('-');
      if (ranges.length == 2) {
        int lower = int.tryParse(ranges.first) ?? -1;
        int higher = int.tryParse(ranges.last) ?? -1;
        if (lower <= higher) {
          return new List.generate(higher - lower + 1, (i) => i + lower);
        }
      }
    }
  }
  throw 'Unable to parse: $constraint';
}

class _CronIterator implements HasNext<DateTime> {
  Schedule _schedule;
  DateTime _currentDate;

  _CronIterator(this._schedule, this._currentDate) {
    _currentDate = DateTime.fromMillisecondsSinceEpoch(this._currentDate.millisecondsSinceEpoch ~/ 60000 * 60000);
  }

  DateTime next() {
    _currentDate = _currentDate.add(Duration(minutes: 1));
    while (true) {
      if (_schedule?.months?.contains(_currentDate.month) == false) {
        _currentDate = DateTime(_currentDate.year, _currentDate.month + 1, 1);
        continue;
      }
      if (_schedule?.weekdays?.contains(_currentDate.weekday) == false) {
        _currentDate = DateTime(_currentDate.year, _currentDate.month, _currentDate.day + 1);
        continue;
      }
      if (_schedule?.days?.contains(_currentDate.day) == false) {
        _currentDate = DateTime(_currentDate.year, _currentDate.month, _currentDate.day + 1);
        continue;
      }
      if (_schedule?.hours?.contains(_currentDate.hour) == false) {
        _currentDate = _currentDate.add(Duration(hours: 1));
        _currentDate = _currentDate.subtract(Duration(minutes: _currentDate.minute));
        continue;
      }
      if (_schedule?.minutes?.contains(_currentDate.minute) == false) {
        _currentDate = _currentDate.add(Duration(minutes: 1));
        continue;
      }
      return _currentDate;
    }
  }
}


example/example.dart
import 'package:cron_parser/cron_parser.dart';
import 'package:timezone/standalone.dart';
import 'package:timezone/timezone.dart';

void main() {
  // by default next cron dates are starting from the current date
  var cronIterator = Cron().parse("0 * * * *", "Europe/London");
  TZDateTime nextDate = cronIterator.next();
  TZDateTime afterNextDate = cronIterator.next();

  // specify a start date to get cron dates after this date
  TZDateTime startDate = TZDateTime(getLocation("Europe/London"), 2020, 4, 01);
  cronIterator = Cron().parse("0 * * * *", "Europe/London", startDate);
  nextDate = cronIterator.next(); // 2020-04-01 01:00:00.000+0100
  var currentDateNext = cronIterator.current(); // 2020-04-01 01:00:00.000+0100
  afterNextDate = cronIterator.next(); // 2020-04-01 02:00:00.000+0100

  // by default previous cron dates are starting from the current date
  var cronIteratorPrevious = Cron().parse("0 * * * *", "Europe/London");
  TZDateTime previousDate = cronIterator.previous();
  TZDateTime beforePreviousDate = cronIterator.previous();

  // specify a start date to get cron dates before this date
  TZDateTime startDatePrevious =
      TZDateTime(getLocation("Europe/London"), 2020, 4, 01);
  cronIterator = Cron().parse("0 * * * *", "Europe/London", startDatePrevious);
  previousDate = cronIterator.previous(); // 2020-03-31 23:00:00.000+0100
  var currentDatePrevious =
      cronIterator.current(); // 2020-03-31 23:00:00.000+0100
  beforePreviousDate = cronIterator.previous(); // 2020-03-31 22:00:00.000+0100
}
